File: src/main/java/mediathek/tool/table/MVTable.java
package mediathek.tool.table;

import mediathek.config.Konstanten;
import mediathek.config.MVConfig;
import mediathek.gui.messages.FontSizeChangedEvent;
import mediathek.tool.MessageBus;
import net.engio.mbassy.listener.Handler;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jspecify.annotations.NonNull;

import javax.swing.*;
import javax.swing.plaf.UIResource;
import javax.swing.table.AbstractTableModel;
import javax.swing.table.TableColumnModel;
import java.awt.*;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public abstract class MVTable extends JTable {
    private static final Logger logger = LogManager.getLogger();
    protected final int[] breite;
    protected final int[] reihe;
    protected final int maxSpalten;
    protected final ColumnVisibilityStore spaltenAnzeigen;
    protected final Optional<MVConfig.Configs> showIconsConfigKey;
    protected final Optional<MVConfig.Configs> smallSenderIconConfigKey;
    /**
     * unmodified JTable used to calculate the row height. Reference only.
     */
    private final JTable probe = new JTable();
    protected boolean useSmallSenderIcons;
    protected List<? extends RowSorter.SortKey> listeSortKeys;
    private int[] selRows = {};
    private boolean showSenderIcon;
    private boolean lineBreak = true;
    protected MVTable(int maxColumns, @NonNull ColumnVisibilityStore visibleColumnStore,
                   @NonNull Optional<MVConfig.Configs> showIconsConfigKey,
                   @NonNull Optional<MVConfig.Configs> smallSenderIconConfigKey) {
        maxSpalten = maxColumns;
        this.showIconsConfigKey = showIconsConfigKey;
        this.smallSenderIconConfigKey = smallSenderIconConfigKey;
        spaltenAnzeigen = visibleColumnStore;
        // make all columns visible by default in column store
        spaltenAnzeigen.fill(true);

        setAutoCreateRowSorter(true);
        setAutoResizeMode(JTable.AUTO_RESIZE_OFF);

        breite = new int[maxSpalten];
        Arrays.fill(breite,-1);

        reihe = new int[maxSpalten];
        Arrays.fill(reihe, -1);

        showIconsConfigKey.ifPresent( key -> showSenderIcon = Boolean.parseBoolean(MVConfig.get(key)));
        smallSenderIconConfigKey.ifPresent(key -> useSmallSenderIcons = Boolean.parseBoolean(MVConfig.get(key)));

        calculateRowHeight();

        MessageBus.getMessageBus().subscribe(this);
    }

    public boolean getUseSmallSenderIcons() {
        return useSmallSenderIcons;
    }

    public void setUseSmallSenderIcons(boolean useSmallSenderIcons) {
        this.useSmallSenderIcons = useSmallSenderIcons;
    }

    protected Color defaultRowBackground(int row) {
        if ((row % 2) != 0) {
            var alternateRowColor = UIManager.getColor("Table.alternateRowColor");
            if (alternateRowColor != null) {
                return alternateRowColor;
            }
        }

        var background = getBackground();
        if (!(background instanceof UIResource)) {
            return background;
        }

        var tableBackground = UIManager.getColor("Table.background");
        return tableBackground != null ? tableBackground : background;
    }

    @Handler
    private void handleFontSizeChanged(FontSizeChangedEvent e) {
        SwingUtilities.invokeLater(this::calculateRowHeight);
    }

    public boolean showSenderIcons() {
        return showSenderIcon;
    }

    public void setShowIcon(boolean newVal) {
        showSenderIcon = newVal;
    }

    public boolean isLineBreak() { return lineBreak;}

    public void setLineBreak(boolean lb) {
        lineBreak = lb;
    }

    /**
     * Return a fictious size of a multi-line text area.
     * @return The fictious size of a multi-line label.
     */
    private int getSizeArea() {
        int lineHeight = getFontMetrics(getFont()).getHeight();
        return lineBreak ? lineHeight * 3 : lineHeight;
    }

    /**
     * Calculate the row height in a table based on icon display,etc.
     */
    public void calculateRowHeight() {
        int minimumHeight = Konstanten.TABLE_DEFAULT_ROW_HEIGHT;

        if (showSenderIcon) {
            minimumHeight = useSmallSenderIcons
                    ? Math.max(Konstanten.TABLE_DEFAULT_ROW_HEIGHT, probe.getRowHeight())
                    : Math.max(Konstanten.TABLE_DEFAULT_LARGE_ICON_ROW_HEIGHT, probe.getRowHeight());
        }

        setRowHeight(Math.max(minimumHeight, getSizeArea()));
    }

    private boolean isColumnVisible(int index) {
        return spaltenAnzeigen.isVisible(index);
    }

    protected void setSpaltenEinAus(int[] nr) {
        for (int i = 0; i < spaltenAnzeigen.length(); ++i) {
            spaltenAnzeigen.setVisible(i, nr[i] > 0);
        }
    }

    public void fireTableDataChanged(boolean setSpalten) {
        if (setSpalten) {
            saveSelectedTableRows();
        }
        var model = (AbstractTableModel)getModel();
        model.fireTableDataChanged();
        if (setSpalten) {
            restoreSelectedTableRows();
        }
    }

    public void scrollToSelection() {
        final int rowCount = getRowCount();

        if (rowCount > 0) {
            int i = getSelectedRow();
            if (i == -1) {
                i = 0;
                getSelectionModel().setSelectionInterval(0, 0);
            }
            if (i >= rowCount) {
                i = rowCount - 1;
            }

            scrollToIndexDelegate(i);
        }
    }

    protected void scrollToIndexDelegate(int index) {
        scrollRectToVisible(getCellRect(index, 0, true));
    }

    protected void saveSelectedTableRows() {
        // Einstellungen der Tabelle merken
        selRows = getSelectedRows();
    }

    protected void restoreSelectedTableRows() {
        if (selRows.length > 0) {
            final int visibleRow;
            if (selRows.length == 1) {
                final var selectedRow = selRows[0];
                selectionModel.setSelectionInterval(selectedRow, selectedRow);
                visibleRow = selectedRow;
            } else {
                selectionModel.setValueIsAdjusting(true);
                for (int selectedRow : selRows) {
                    if (selectedRow < getRowCount()) {
                        addRowSelectionInterval(selectedRow, selectedRow);
                    }
                }
                selectionModel.setValueIsAdjusting(false);
                visibleRow = selRows[0];
            }
            scrollToIndexDelegate(visibleRow);
            requestFocusInWindow();
        }
    }

    protected void changeTableModelColumnWidths() {
        final TableColumnModel model = getColumnModel();
        for (int i = 0; i < breite.length && i < getColumnCount(); ++i) {
            final int colIndex = convertColumnIndexToView(i);
            var column = model.getColumn(colIndex);
            if (breite[i] == 0) {
                column.setMinWidth(0);
                column.setPreferredWidth(0);
                column.setMaxWidth(0);
            } else {
                column.setMinWidth(10);
                column.setMaxWidth(3000);
                column.setPreferredWidth(breite[i]);
            }
        }
    }

    protected void changeInternalColumnWidths() {
        for (int i = 0; i < breite.length && i < getColumnCount(); ++i) {
            if (!isColumnVisible(i)) {
                // geänderte Ansicht der Spalten abfragen
                breite[i] = 0;
            } else if (breite[i] == 0) {
                breite[i] = 100; // damit sie auch zu sehen ist :)
            }
        }
    }

    public void spaltenEinAus() {
        getSpalten(); // die aktuelle Breite holen
        changeInternalColumnWidths();
        changeTableModelColumnWidths();

        validate();
    }

    public void getSpalten() {
        // Einstellungen der Tabelle merken
        saveSelectedTableRows();

        var columnCount = getModel().getColumnCount();

        for (int i = 0; i < reihe.length && i < columnCount; ++i) {
            reihe[i] = convertColumnIndexToModel(i);
        }

        for (int i = 0; i < breite.length && i < columnCount; ++i) {
            breite[i] = getColumnModel().getColumn(convertColumnIndexToView(i)).getWidth();
        }
        if (this.getRowSorter() != null) {
            listeSortKeys = getRowSorter().getSortKeys();
        } else {
            listeSortKeys = null;
        }
    }

    public void setSpalten() {
        // gemerkte Einstellungen der Tabelle wieder setzen
        try {
            changeInternalColumnWidths();

            final TableColumnModel model = getColumnModel();
            changeTableModelColumnWidths();

            for (int i = 0; i < reihe.length && i < getColumnCount(); ++i) {
                model.moveColumn(convertColumnIndexToView(reihe[i]), i);
            }

            if (listeSortKeys != null) {
                if (!listeSortKeys.isEmpty()) {
                    getRowSorter().setSortKeys(listeSortKeys);
                }
            }

            restoreSelectedTableRows();

            validate();
        } catch (Exception ex) {
            logger.error("setSpalten", ex);
        }
    }

    /**
     * Perform common reset steps for all subclasses.
     */
    public void resetTabelle() {
        listeSortKeys = null;

        if (getRowSorter() != null) {
            getRowSorter().setSortKeys(null); // empty sort keys
        }
        setRowSorter(null);
        setAutoCreateRowSorter(true);
        spaltenAusschalten();
        setSpaltenEinAus(breite);
        setSpalten();
        calculateRowHeight();
    }

    protected abstract void spaltenAusschalten();

    /**
     * Write table display preferences to config.
     */
    public void writeTableConfigurationData() {
        showIconsConfigKey.ifPresent(key -> MVConfig.add(key, String.valueOf(showSenderIcon)));
        smallSenderIconConfigKey.ifPresent(key -> MVConfig.add(key, String.valueOf(useSmallSenderIcons)));
    }
}

File: src/main/kotlin/mediathek/tool/table/MVTable.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.tool.table

import mediathek.config.Konstanten
import mediathek.config.MVConfig
import mediathek.gui.messages.FontSizeChangedEvent
import mediathek.tool.MessageBus
import net.engio.mbassy.listener.Handler
import org.apache.logging.log4j.LogManager
import java.awt.Color
import javax.swing.JTable
import javax.swing.RowSorter
import javax.swing.SwingUtilities
import javax.swing.UIManager
import javax.swing.plaf.UIResource
import javax.swing.table.AbstractTableModel
import javax.swing.table.TableColumnModel

abstract class MVTable protected constructor(
    @JvmField protected val maxSpalten: Int,
    @JvmField protected val spaltenAnzeigen: ColumnVisibilityStore,
    protected val showIconsConfigKey: java.util.Optional<MVConfig.Configs>,
    protected val smallSenderIconConfigKey: java.util.Optional<MVConfig.Configs>,
) : JTable() {

    @JvmField
    protected val breite = IntArray(maxSpalten) { -1 }

    @JvmField
    protected val reihe = IntArray(maxSpalten) { -1 }

    private var useSmallSenderIconsState = false
    protected var listeSortKeys: List<out RowSorter.SortKey>? = null

    /**
     * Unmodified JTable used to calculate the row height. Reference only.
     */
    private val probe = JTable()
    private var selectedRowsSnapshot = IntArray(0)
    private var showSenderIcon = false
    private var lineBreak = true

    init {
        spaltenAnzeigen.fill(true)

        autoCreateRowSorter = true
        autoResizeMode = AUTO_RESIZE_OFF

        showIconsConfigKey.ifPresent { showSenderIcon = MVConfig.get(it).toBoolean() }
        smallSenderIconConfigKey.ifPresent { useSmallSenderIconsState = MVConfig.get(it).toBoolean() }

        calculateRowHeight()
        MessageBus.messageBus.subscribe(this)
    }

    fun getUseSmallSenderIcons(): Boolean = useSmallSenderIconsState

    fun setUseSmallSenderIcons(useSmallSenderIcons: Boolean) {
        useSmallSenderIconsState = useSmallSenderIcons
    }

    protected fun defaultRowBackground(row: Int): Color {
        if (row % 2 != 0) {
            UIManager.getColor("Table.alternateRowColor")?.let { return it }
        }

        val background = background
        if (background !is UIResource) {
            return background
        }

        return UIManager.getColor("Table.background") ?: background
    }

    @Suppress("UNUSED_PARAMETER")
    @Handler
    private fun handleFontSizeChanged(event: FontSizeChangedEvent) {
        SwingUtilities.invokeLater(::calculateRowHeight)
    }

    fun showSenderIcons(): Boolean = showSenderIcon

    fun setShowIcon(newVal: Boolean) {
        showSenderIcon = newVal
    }

    fun isLineBreak(): Boolean = lineBreak

    fun setLineBreak(lineBreak: Boolean) {
        this.lineBreak = lineBreak
    }

    /**
     * Return a fictious size of a multi-line text area.
     */
    private fun sizeArea(): Int {
        val lineHeight = getFontMetrics(font).height
        return if (lineBreak) {
            lineHeight * 3
        } else {
            lineHeight
        }
    }

    /**
     * Calculate the row height in a table based on icon display, etc.
     */
    open fun calculateRowHeight() {
        val minimumHeight = if (showSenderIcon) {
            if (useSmallSenderIconsState) {
                maxOf(Konstanten.TABLE_DEFAULT_ROW_HEIGHT, probe.rowHeight)
            } else {
                maxOf(Konstanten.TABLE_DEFAULT_LARGE_ICON_ROW_HEIGHT, probe.rowHeight)
            }
        } else {
            Konstanten.TABLE_DEFAULT_ROW_HEIGHT
        }

        rowHeight = maxOf(minimumHeight, sizeArea())
    }

    private fun isColumnVisible(index: Int): Boolean = spaltenAnzeigen.isVisible(index)

    protected fun setSpaltenEinAus(columnWidths: IntArray) {
        for (index in 0 until spaltenAnzeigen.length()) {
            spaltenAnzeigen.setVisible(index, columnWidths[index] > 0)
        }
    }

    fun fireTableDataChanged(setSpalten: Boolean) {
        if (setSpalten) {
            saveSelectedTableRows()
        }

        (model as AbstractTableModel).fireTableDataChanged()

        if (setSpalten) {
            restoreSelectedTableRows()
        }
    }

    fun scrollToSelection() {
        val rowCount = rowCount
        if (rowCount <= 0) {
            return
        }

        var selectedRow = selectedRow
        if (selectedRow == -1) {
            selectedRow = 0
            selectionModel.setSelectionInterval(0, 0)
        }
        if (selectedRow >= rowCount) {
            selectedRow = rowCount - 1
        }

        scrollToIndexDelegate(selectedRow)
    }

    protected fun scrollToIndexDelegate(index: Int) {
        scrollRectToVisible(getCellRect(index, 0, true))
    }

    protected open fun saveSelectedTableRows() {
        selectedRowsSnapshot = selectedRows
    }

    protected open fun restoreSelectedTableRows() {
        if (selectedRowsSnapshot.isEmpty()) {
            return
        }

        val visibleRow = if (selectedRowsSnapshot.size == 1) {
            val selectedRow = selectedRowsSnapshot[0]
            selectionModel.setSelectionInterval(selectedRow, selectedRow)
            selectedRow
        } else {
            selectionModel.valueIsAdjusting = true
            try {
                for (selectedRow in selectedRowsSnapshot) {
                    if (selectedRow < rowCount) {
                        addRowSelectionInterval(selectedRow, selectedRow)
                    }
                }
            } finally {
                selectionModel.valueIsAdjusting = false
            }
            selectedRowsSnapshot[0]
        }

        scrollToIndexDelegate(visibleRow)
        requestFocusInWindow()
    }

    protected fun changeTableModelColumnWidths() {
        val tableColumnModel: TableColumnModel = columnModel
        for (index in 0 until minOf(breite.size, columnCount)) {
            val column = tableColumnModel.getColumn(convertColumnIndexToView(index))
            if (breite[index] == 0) {
                column.minWidth = 0
                column.preferredWidth = 0
                column.maxWidth = 0
            } else {
                column.minWidth = 10
                column.maxWidth = 3000
                column.preferredWidth = breite[index]
            }
        }
    }

    protected fun changeInternalColumnWidths() {
        for (index in 0 until minOf(breite.size, columnCount)) {
            when {
                !isColumnVisible(index) -> breite[index] = 0
                breite[index] == 0 -> breite[index] = 100
            }
        }
    }

    fun spaltenEinAus() {
        getSpalten()
        changeInternalColumnWidths()
        changeTableModelColumnWidths()
        validate()
    }

    open fun getSpalten() {
        saveSelectedTableRows()

        val modelColumnCount = model.columnCount
        for (index in 0 until minOf(reihe.size, modelColumnCount)) {
            reihe[index] = convertColumnIndexToModel(index)
        }

        val tableColumnModel = columnModel
        for (index in 0 until minOf(breite.size, modelColumnCount)) {
            breite[index] = tableColumnModel.getColumn(convertColumnIndexToView(index)).width
        }

        listeSortKeys = rowSorter?.sortKeys
    }

    open fun setSpalten() {
        try {
            changeInternalColumnWidths()

            val tableColumnModel: TableColumnModel = columnModel
            changeTableModelColumnWidths()

            for (index in 0 until minOf(reihe.size, columnCount)) {
                tableColumnModel.moveColumn(convertColumnIndexToView(reihe[index]), index)
            }

            val savedSortKeys = listeSortKeys
            if (!savedSortKeys.isNullOrEmpty()) {
                rowSorter?.sortKeys = savedSortKeys
            }

            restoreSelectedTableRows()
            validate()
        } catch (exception: Exception) {
            logger.error("setSpalten", exception)
        }
    }

    /**
     * Perform common reset steps for all subclasses.
     */
    open fun resetTabelle() {
        listeSortKeys = null

        rowSorter?.sortKeys = null
        rowSorter = null
        autoCreateRowSorter = true
        spaltenAusschalten()
        setSpaltenEinAus(breite)
        setSpalten()
        calculateRowHeight()
    }

    protected abstract fun spaltenAusschalten()

    /**
     * Write table display preferences to config.
     */
    open fun writeTableConfigurationData() {
        showIconsConfigKey.ifPresent { MVConfig.add(it, showSenderIcon.toString()) }
        smallSenderIconConfigKey.ifPresent { MVConfig.add(it, useSmallSenderIconsState.toString()) }
    }

    companion object {
        private val logger = LogManager.getLogger()
    }
}
